sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/ContinuousDataReader.java

```diff
@@ -0,0 +1,30 @@
+package org.apache.spark.sql.sources.v2.reader;
+
+import org.apache.spark.sql.sources.v2.reader.PartitionOffset;
+
+import java.io.IOException;
+
+/**
+ * A variation on {@link DataReader} for use with streaming in continuous processing mode.
+ */
+public interface ContinuousDataReader<T> extends DataReader<T> {
+    /**
+     * Proceed to next record, returning false only if the read is interrupted.
+     *
+     * @throws IOException if failure happens during disk/network IO like reading files.
+     */
+    boolean next() throws IOException;
+
+    /**
+     * Return the current record. This method should return same value until `next` is called.
+     */
+    T get();
+
+    /**
+     * Get the offset of the next record; that is, the earliest record after the current result of
+     * {@link this.get}.
+     *
+     * The execution engine will use this offset as a restart checkpoint.
+     */
+    PartitionOffset getOffset();
```

Comment:
Maybe this should be `getNextOffset()` to be more explicit? When I see `getX` I automatically think current value of X, not the next value of X.

Reply:
I think that this should be the offset of the current record, right?  Source like kinesis do not give you the next offset and typically our ranges are `(start, end]` (i.e. exclusive for the start).

Reply:
That explains a lot of why I was finding offset manipulation awkward. Yes, if the ranges are (start, end] this should be the current offset.

Reply:
Agreed.